ticketbuyer/purchase.go

```diff
@@ -48,14 +51,20 @@ out:
 	for {
 		select {
 		case v := <-p.ntfnChan:
+			p.wg.Add(1)
 			go func(s1, s2 chan struct{}) {
-				<-s1 // wait for previous worker to finish
+				select {
+				case <-s1: // wait for previous worker to finish
 
-				// Purchase tickets for each attached block, not just for the
-				// update to the main chain.  This is probably not optimal but
-				// it matches how dcrticketbuyer worked.
-				for h := v.NewHeight - int32(len(v.AttachedBlocks)) + 1; h <= v.NewHeight; h++ {
-					p.purchase(int64(h))
+					// Purchase tickets for each attached block, not just for the
+					// update to the main chain.  This is probably not optimal but
+					// it matches how dcrticketbuyer worked.
+					for h := v.NewHeight - int32(len(v.AttachedBlocks)) + 1; h <= v.NewHeight; h++ {
+						p.purchase(int64(h))
+					}
+					p.wg.Done()
+				case <-p.quit:
+					return
```

Comment:
Let me check that. I thought all workers which are blocked on select would return receive quit and return early.